amethyst_error/src/lib.rs

```diff
@@ -0,0 +1,402 @@
+//! Contains the `Error` type and company as used by Amethyst.
+//!
+//! **Note:** This type is not intended to be used outside of Amethyst.
+//! If you are integrating a crate from amethyst to use this, it is recommended that you treat this
+//! type as an opaque [`std::error::Error`].
+//!
+//! [`std::error::Error`]: https://doc.rust-lang.org/std/error/trait.Error.html
+use self::internal::{Backtrace, HAS_BACKTRACE};
+use std::borrow::Cow;
+use std::env;
+use std::error;
+use std::ffi;
+use std::fmt;
+use std::result;
+use std::sync::atomic;
+
+/// Private
+enum ErrorImpl {
+    Message(Cow<'static, str>),
+    Error(Box<dyn error::Error + Send + Sync>),
```

Comment:
Current `ametyst::Error` is not `Sync`, why requiring it here?

Reply:
Accidentally resolved. But it makes life easier, and I'm mimicking the behavior of failure.

Reply:
Working around `!Sync` is a lot harder than `Sync`. Enforcing `Sync` is IMO the more conservative option, and I didn't encounter any instance in the current codebase where we needed `!Sync`.

In the future we can loosen the constraint if we find a good case for it. But I'd prefer to wait until we have one.

Reply:
This has also kinda become the norm in error types now because of the popularity of failure.

Reply:
I kinda agree, I would like to drop the `Sync` constraint. I think it will cause way more pain then it will benefit.

Reply:
Is there real use case for writing `!Sync` error type?

Reply:
From the nomicron:

> Major exceptions include:
>
>  - raw pointers are neither Send nor Sync (because they have no safety guards).
>  - UnsafeCell isn't Sync (and therefore Cell and RefCell aren't).
>  - Rc isn't Send or Sync (because the refcount is shared and unsynchronized).

So the error type would have to have some kind of shared ownership, basically one of `RefCell., `Cell`, `Rc`, or `*(const|mut)`. I'd be very suprised if this happens in practice, and I think it would denote a design error - `std::error::Error` is when something's gone wrong and you're probably just going to print it to the user and give up (it implements `Display` and has a `source`) - if you're doing actual game logic in it use a custom type.